### PR TITLE
viewer#1200 Avatar rotates 360 degrees when viewed from the top and below

### PR DESCRIPTION
**Pull request by LLGuru**
See https://github.com/secondlife/viewer/issues/1200

This change applies `pitch()` to a copy of the `mFrameAgent` and checks the resulting vertical angle
The new value of `mFrameAgent` is used only if the resulting vertical angle doesn't exceed the limits

This method is more precise than any attempts to guess the resulting angle before applying `pitch()`

### CODE REVIEW

File: indra/newview/llagent.cpp [hunk 1]
```diff
@@ -1453,48 +1453,70 @@ LLVector3 LLAgent::getReferenceUpVector()
 	return up_vector;
 }
 
-
+#pragma optimize("", off)
```
Comment (marchcat) [inline indra/newview/llagent.cpp:1456]: ?

Comment (LLGuru) [reply indra/newview/llagent.cpp]: > ?

I felt that I would forget to remove it :-)

**Review by marchcat — APPROVED** [review]
Let's test it.

File: indra/newview/llagentcamera.h [hunk 2]
```diff
@@ -112,6 +112,7 @@ class LLAgentCamera
 	//--------------------------------------------------------------------
 public:
 	void switchCameraPreset(ECameraPreset preset);
+	ECameraPreset getCameraPreset() const { return mCameraPreset; }
```
Comment (akleshchev) [inline indra/newview/llagentcamera.h:115]: Why is this needed?

File: indra/newview/llagent.cpp [hunk 3]
```diff
@@ -1453,48 +1453,69 @@ LLVector3 LLAgent::getReferenceUpVector()
...
-		LLVector3 agent_focus_pos = getPosAgentFromGlobal(gAgentCamera.calcFocusPositionTargetGlobal());
-		LLVector3 look_dir = agent_focus_pos - viewer_camera_pos;
-		F32 angle_from_skyward = angle_between(look_dir, skyward);
-		if (angle_from_skyward + angle < look_up_limit)
-		{
-			angle = look_up_limit - angle_from_skyward;
-		}
-	}
+    // don't let user pitch if pointed almost all the way down or up
+    if (gAgentCamera.getCameraMode() == CAMERA_MODE_THIRD_PERSON ||
```
Comment (akleshchev) [inline indra/newview/llagent.cpp:1474]: Why did yu add this cheack? Won't this negatively affect avatar customization?

Comment (marchcat) [reply indra/newview/llagent.cpp:1474]: Good point.

File: indra/newview/llagentcamera.h (hunk 2 above)
Comment (LLGuru) [reply indra/newview/llagentcamera.h:115]: See the second file in this commit

File: indra/newview/llagent.cpp (hunk 3 above)
Comment (LLGuru) [reply indra/newview/llagent.cpp:1474]: Because each preset requires different checks

Comment (LLGuru) [reply indra/newview/llagent.cpp:1474]: The avatar customization mode won't be affected because it doesn't use vertical angles close to 1 and 179

File: indra/newview/llagentcamera.h (hunk 2 above)
Comment (akleshchev) [reply indra/newview/llagentcamera.h:115]: And what would happen with custom presets?

File: indra/newview/llagent.cpp (hunk 3 above)
Comment (akleshchev) [reply indra/newview/llagent.cpp:1474]: And CAMERA_MODE_FOLLOW? That's a mode where script overrides control over camera and otherwise acts as a CAMERA_MODE_THIRD_PERSON, doesn't appear to affect mouselook

File: indra/newview/llagentcamera.h (hunk 2 above)
Comment (maxim-productengine) [reply indra/newview/llagentcamera.h:115]: I just want to mention that all 3 "default" presets (Front, Rear, Side) can be overwritten. For example, you can set the values of 'Rear' and 'Side' to be the same as 'Front'. Basically, they are almost the same as 'custom' presets, but have their own buttons and default values, stored in the file.

File: indra/newview/llagent.cpp (hunk 3 above)
Comment (akleshchev) [reply indra/newview/llagent.cpp]: I checked and it behaves exactly like CAMERA_MODE_THIRD_PERSON

File: indra/newview/llagentcamera.h (hunk 2 above)
Comment (LLGuru) [reply indra/newview/llagentcamera.h:115]: > So in this case, it looks like those bugs will appear in the very first round of testing. Why wait, then?

Nothing is changed for custom presets
If we don't have bugs with them right now, why should they appear after this change?

File: indra/newview/llagent.cpp (hunk 3 above)
Comment (LLGuru) [reply indra/newview/llagent.cpp]: With CAMERA_MODE_THIRD_PERSON your camera could rotate any directions without any action from your side
So this bug doesn't make sense for CAMERA_MODE_THIRD_PERSON

When the vertical angle takes critical values 1 or 179, camera could rotate, and nobody treats this like a bug

File: indra/newview/llagentcamera.h (hunk 2 above)
Comment (maxim-productengine) [reply indra/newview/llagentcamera.h:115]: I meant something like that:
- Open 'Camer controls'
- Select 'Front view'
- Click 'Save as preset' button
- Choose 'Replace a preset' and select 'Rear'
- (then repeat and select 'Side')
- Click 'Save as preset' button once again
- And save as new preset.

After that the bug will be reproducible on all camera preset views, including REAR, GROUP, CUSTOM.

Comment (akleshchev) [reply indra/newview/llagentcamera.h:115]: Test plan for QA now includes testing presets, so I'm tentatively approving this.

**Review by akleshchev — APPROVED** [review]

**Comment by akleshchev** [comment]
But may be retarget it to maint C?

**Comment by marchcat** [comment]
> retarget it to maint C?

Sounds good.

File: indra/newview/llagentcamera.h (hunk 2 above)
Comment (LLGuru) [reply indra/newview/llagentcamera.h:115]: If the user explicitly sets incorrect preset settings, the bugs like this are expected, but they won't be treated as bugs

For example, if user choses the color scheme with white color for font and background, the text becomes unreadable, but this is not a buy, yes? :-)

**Review by marchcat — APPROVED** [review]
